cli_tools/common/utils/daisyutils/daisy_utils.go

```diff
@@ -355,19 +355,27 @@ func PostProcessDErrorForNetworkFlag(action string, err error, network string, w
 	}
 }
 
-// RunWorkflowWithCancelSignal runs Daisy workflow with accepting Ctrl-C signal
-func RunWorkflowWithCancelSignal(ctx context.Context, w *daisy.Workflow) error {
+// RunWorkflowWithCancelSignal runs a Daisy workflow, and allows for cancellation from two sources:
+//   1. The user types Ctrl-C on their keyboard.
+//   2. The caller sends a cancellation reason on the cancel channel (or closes it).
+func RunWorkflowWithCancelSignal(w *daisy.Workflow, cancel <-chan string) error {
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, os.Interrupt)
 	go func(w *daisy.Workflow) {
 		select {
+		case reason := <-cancel:
+			if reason != "" {
+				w.CancelWithReason(reason)
+			}
```

Comment:
What is reason is not provided? Should we call `w.CancelWorkflow()`?

Reply:
In the current implementation, the string is empty when the channel is closed. That said, I can see how that would be confusing, both now and in the future. So I added a call to `w.CancelWorkflow()`, which is safe since we added a guard that allows daisy workflow cancellation to be idempotent.